### PR TITLE
connects AgileVentures/MetPlus_tracker#200

### PR DESCRIPTION
**Pull request by amba178**
I have change the Job<>Address relation in the model. We need to updated our ERD, I think.

### CODE REVIEW

File: app/controllers/jobs_controller.rb [hunk 1]
```diff
@@ -13,11 +15,14 @@ def index
 	end	
 
 	def new
-		@job = Job.new(:company_id => params[:company_id])
+		@job = Job.new(:company_id => params[:company_id],
+			           :company_person_id => params[:company_person_id])
```
Comment (joaopapereira) [inline app/controllers/jobs_controller.rb]: Can't a JD create a job?

File: app/controllers/jobs_controller.rb [hunk 2]
```diff
@@ -13,11 +15,14 @@ def index
...
 	end 
 
 	def create
 		@job = Job.new(job_params)
+		 
+		@job.address = @company_person.address
```
Comment (joaopapereira) [inline app/controllers/jobs_controller.rb]: Is this assumption something that Chet asked for? A company person can only create a job at his address?

File: app/controllers/jobs_controller.rb [hunk 3]
```diff
@@ -51,9 +58,12 @@ def destroy
 
 		def authentication_for_post_or_edit 
 			if !company_person?
-			   flash[:alert] = "Sorry, You are not allowed to post or edit a job!"
```
Comment (joaopapereira) [inline app/controllers/jobs_controller.rb]: Revert this change, we don't need spaces there

File: app/controllers/jobs_controller.rb [hunk 4]
```diff
@@ -69,6 +79,14 @@ def company_person?
 			end
 		end
 
+		def is_company_person_job
+			 set_company_person
+			 if !(@job.address == @company_person.address)
```
Comment (joaopapereira) [inline app/controllers/jobs_controller.rb]: Not sure this if it's accurate

File: app/controllers/jobs_controller.rb (hunk 1 above)
Comment (amba178) [inline app/controllers/jobs_controller.rb]: From my understanding JD work for agency, and we know that has some kind of ties with company that is help to match jobs with jobseekers, but  he/she not permitted to create a job, same applies to case manager. What do you think?

File: app/controllers/jobs_controller.rb (hunk 3 above)
Comment (amba178) [inline app/controllers/jobs_controller.rb]: Ok.

File: app/controllers/jobs_controller.rb (hunk 2 above)
Comment (amba178) [inline app/controllers/jobs_controller.rb]: I think so, walmart A  jobs can not be created, edited or delete by walmart B company_person vise verser. Otherwise there will be some kind of confusion. Can you send me Chet's email through slack private messaging/direct messaging if you can. I have some questions to ask.

File: app/views/layouts/_footer.html.haml [hunk 5]
```diff
@@ -9,7 +9,7 @@
 		%div{class: "row text-center", id: "footer-text-color"}
 			-if pets_user.is_a?(CompanyPerson)
 				For employers:
-				=link_to "Post jobs", new_job_path(company_id: current_user.try(:actable).try(:company_id))
+				=link_to "Post jobs", new_job_path(company_id: current_user.try(:actable).try(:company_id), company_person_id: current_user.try(:actable_id))
 				|
```
Comment (joaopapereira) [inline app/views/layouts/_footer.html.haml]: Does it make sense to show this link if you cannot post jobs?

File: app/views/layouts/application.html.haml [hunk 6]
```diff
@@ -40,7 +40,7 @@
             - @model_errors.full_messages.each do |msg|
               %li= msg
       %div{:class => 'row'}
-        = yield
+        = yield
```
Comment (joaopapereira) [inline app/views/layouts/application.html.haml]: We don't need this space

File: db/migrate/20151107120000_create_database.rb [hunk 7]
```diff
@@ -4,8 +4,7 @@ def change
       t.string :street
       t.string :city
       t.string :zipcode
-      t.references :location, polymorphic: true, index: true
-
+      t.references :location, polymorphic: true, index: true
```
Comment (joaopapereira) [inline db/migrate/20151107120000_create_database.rb]: This is not a good practice. If you need to change a table do a new migration. Before we go live we will merge all changes but in the meanwhile create new migrations

File: spec/controllers/jobs_controller_spec.rb [hunk 8]
```diff
@@ -249,17 +247,21 @@
...
                              company_id: '3',
-                             :shift => "Evening", company_job_id: "WERRR123"}
+                             address_id: '2',
+                             company_person_id: 3,
+                             :shift => "Evening", 
+                             company_job_id: "WERRR123"}
       expect(response).to redirect_to(:action => 'index')
       should set_flash 
-      expect(Job.count).to eq(2)
+      expect(Job.count).to eq(3)
```
Comment (joaopapereira) [inline spec/controllers/jobs_controller_spec.rb]: This code doesn't look good. You should get the count before the post and then compare after

File: spec/factories/jobs.rb [hunk 9]
```diff
@@ -3,8 +3,8 @@
     title "MyString"
     description "MyString"
     company_id 1
-    # company_person_id 1
-    # address
+    company_person_id 1
```
Comment (joaopapereira) [inline spec/factories/jobs.rb]: Shouldn't we call the factories of company and address here?

**Comment by joaopapereira** [comment]
Looks good please review my comments

File: app/views/layouts/_footer.html.haml (hunk 5 above)
Comment (amba178) [inline app/views/layouts/_footer.html.haml]: Only shows if you are company_person instance

File: app/views/layouts/application.html.haml (hunk 6 above)
Comment (amba178) [inline app/views/layouts/application.html.haml]: ok

File: db/migrate/20151107120000_create_database.rb (hunk 7 above)
Comment (amba178) [inline db/migrate/20151107120000_create_database.rb]: you are right. But here I do not why there are two return lines, and will be fixed.

File: spec/controllers/jobs_controller_spec.rb (hunk 8 above)
Comment (amba178) [inline spec/controllers/jobs_controller_spec.rb]: yes. Good suggestion. thanks

File: spec/factories/jobs.rb (hunk 9 above)
Comment (amba178) [inline spec/factories/jobs.rb]: actually, should be foreign key for address, instead of instance of address.

File: app/controllers/jobs_controller.rb (hunk 1 above)
Comment (patmbolger) [inline app/controllers/jobs_controller.rb]: This is something we need to check with the client on.

The client wrote this:

```
 I want a standard format that employers and job developers can input job leads in which will be recognizable by the PETS software.
```
in this document: http://www.agileventures.org/projects/metplus/documents/public-employment-tracking-system-description-1

Then, at the start of the project we wrote this requirement for the interaction between job developers and employers:
```
JD:E/3	JD [job developer] creates a job opportunity from the E [employer] description
```
in this document: http://www.agileventures.org/projects/metplus/documents/roles-and-interactions

File: app/controllers/jobs_controller.rb (hunk 2 above)
Comment (patmbolger) [inline app/controllers/jobs_controller.rb]: I think this is way too restrictive.

A company can have many addresses - reflecting locations where the company may have available jobs.

A company person has an address.  We don't restrict that address to being in the set of company addresses, nor should we necessarily do that.  IMO, this is for assistance in contacting a company person if we need to.

A job address might be restricted to being one of the company addresses - that seems reasonable to me.

However, this code would prevent an HR company person (whose address is the company's HQ, or their home if they are a remote employee, or anywhere really) from posting jobs for other locations.

File: app/models/address.rb [hunk 10]
```diff
@@ -1,6 +1,6 @@
 class Address < ActiveRecord::Base
   belongs_to :location, polymorphic: true
-  
+  has_many :jobs
```
Comment (patmbolger) [inline app/models/address.rb]: yes, we need to change the ERD to reflect this.  I will do that.

Also, while you're at it you may want to add
```
has_many :company_people
```
which has been in the ERD but not reflected in the Address model yet.

**Comment by patmbolger** [comment]
My review complete - please see comments.

File: app/models/address.rb (hunk 10 above)
Comment (amba178) [inline app/models/address.rb]: I will. Let set PP time just to work on ERD I think we might need to change or add associations on other models too that way we are on the same page with each other.

File: app/controllers/jobs_controller.rb (hunk 3 above)
Comment (amba178) [inline app/controllers/jobs_controller.rb]: it shows changes on flash message , I think. not spaces.
